src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1075,7 +1075,28 @@ protected BoundExpression BindInferredVariableInitializer(DiagnosticBag diagnost
             }
 
             diagnostics.AddRangeAndFree(localDiagnostics);
-            var boundDeclType = new BoundTypeExpression(typeSyntax, aliasOpt, inferredType: isVar, type: declTypeOpt.TypeSymbol);
+
+            ImmutableArray<BoundExpression> dimensionsOpt;
+            if (typeSyntax is ArrayTypeSyntax arrayTypeSyntax && arrayTypeSyntax.RankSpecifiers.Count > 0)
+            {
+                var sizes = ArrayBuilder<BoundExpression>.GetInstance();
+                // We don't want to report any errors from this, as the syntax is invalid anyway
+                var dummyDiagnostics = DiagnosticBag.GetInstance();
+                foreach (var expressionSyntax in arrayTypeSyntax.RankSpecifiers[0].Sizes)
+                {
+                    var size = BindArrayRankSpecifier(expressionSyntax, typeSyntax, dummyDiagnostics);
```

Comment:
>var size = BindArrayRankSpecifier(expressionSyntax, typeSyntax, dummyDiagnostics); [](start = 20, length = 82)

It looks like, when a declaration statement contains multiple declarators, we are going to bind sizes multiple times. Moreover, the bound tree is going to have multiple distinct bound nodes associated with the same syntax. This is likely to create problems with SemanticModel and probably in other components. I think the sizes should be bound once (where the typeSyntax is bound) and the bound tree shouldn't contain duplicate nodes associated with sizes.

Reply:
I don't think we do, and I have tested for that in the semantic model, but I will take another look when I get the chance.

What I believe happens is that multiple declarations all have access to the same BoundTypeExpression. There is no obvious way to avoid this. In the Semantic model we just take the first one.

Reply:
Thinking back to it, a Multiple Local Declaration just contains a list of boundLocalDeclarations, and no BoundTypeExpression. Unless we plan to change that, the current approach is I think most sensible.

We already repeat the BoundTypeExpression multiple times in the tree, and the dimensions is a member of the BoundTypeExpression, so I do not see that this is any worse than what we already have.

Reply:
> We already repeat the BoundTypeExpression multiple times in the tree, and the dimensions is a member of the BoundTypeExpression, so I do not see that this is any worse than what we already have.

Yes, we probably have duplicates for BoundTypeExpression today, this is not good and we'd rather not make problem worse by duplicating even more nodes. SemanticModel builds a map from a syntax node to bound nodes associated with the syntax. In general, it is not Ok to have duplicates in that list of associated bound nodes. This might not cause observable problems with BoundTypeExpression because those are not real expressions. Or problems exist, but we simply not aware of them. I am not comfortable with duplicating bound nodes for expressions, because a lot more things happen for regular expressions.

In any case, It is not OK to bind the same expression over and over again when we can easily avoid that. 

---
In reply to: [251624426](https://github.com/dotnet/roslyn/pull/32843#discussion_r251624426) [](ancestors = 251624426)

Reply:
I'll give it a try when I get a chance.

Reply:
@AlekseyTs 
> I think the sizes should be bound once (where the typeSyntax is bound) and the bound tree shouldn't contain duplicate nodes associated with sizes.

Where exactly do you mean by this?

As far as I can see there is no single location where we Bind a Type Syntax. Instead we create a BoundTypeExpression in a number of places, including here.

As far as I can see in a variable declaration the only time the TypeSyntax is Bound is when Binding the BoundLocalDeclaration.

Since a BoundMultipleLocalDeclarations is just a list of of BoundlocalDeclarations, there is no way that I can see to avoid this being bound multiple times.

Perhaps the real problem is with our definition of a BoundMultipleLocalDeclaration and BoundlocalDeclaration.

```xml
  <!--
  Bound node that represents a single local declaration:
    int x = Foo();
    
  NOTE: The node does NOT introduce the referenced local into surrounding scope.
        A local must be explicitly declared in a BoundBlock to be usable inside it.
        
  NOTE: In an error recovery scenario we might have a local declaration parsed as
        int x[123] - This is an error commonly made by C++ developers who come to
        C#. We will give a good error about it at parse time, but we should preserve
        the semantic analysis of the argument list in the bound tree.
  -->
  <Node Name="BoundLocalDeclaration" Base="BoundStatement">
    <Field Name="LocalSymbol" Type="LocalSymbol"/>
    <Field Name="DeclaredType" Type="BoundTypeExpression"/>
    <Field Name="InitializerOpt" Type="BoundExpression" Null="allow"/>
    <Field Name="ArgumentsOpt" Type="ImmutableArray&lt;BoundExpression&gt;" Null="allow"/>
  </Node>

  <!--
  Bound node that represents multiple local declarations:
    int x =1, y =2;
    
  Works like multiple BoundLocalDeclaration nodes.  
  -->
  <Node Name="BoundMultipleLocalDeclarations" Base="BoundStatement">
    <Field Name="LocalDeclarations" Type="ImmutableArray&lt;BoundLocalDeclaration&gt;"/>
  </Node>
```

```csharp
            if (variableCount == 1)
            {
                return BindVariableDeclaration(kind, isVar, variableList[0], typeSyntax, declType, alias, diagnostics, node);
            }
            else
            {
                BoundLocalDeclaration[] boundDeclarations = new BoundLocalDeclaration[variableCount];
                int i = 0;
                foreach (var variableDeclarationSyntax in variableList)
                {
                    boundDeclarations[i++] = BindVariableDeclaration(kind, isVar, variableDeclarationSyntax, typeSyntax, declType, alias, diagnostics);
                }
                return new BoundMultipleLocalDeclarations(node, boundDeclarations.AsImmutableOrNull());
            }
```

We are making a fundamental distinction in the tree between a Declaration of multiple locals and a declaration of only one. When we declare multiple locals we create a list of local declarations, and when we are declaring only one we bind the single declaration directly. I don't see why such a distinction should exist.

Instead every declaration can contain an arbitrary number of locals. There is no need to special case the situation where it contains only one. This means that every single declaration should be a BoundMultipleLocalDeclarations, and every single BoundLocalDeclaration is just an item in the list of BoundMultipleLocalDeclarations. This means we can move DeclaredType to the containing BoundMultipleLocalDeclarations, which will avoid it being duplicated in the tree.

If you agree this should be done, I will obviously do so in a seperate PR as part of a seperate issue.

Reply:
> This means we can move DeclaredType to the containing BoundMultipleLocalDeclarations, which will avoid it being duplicated in the tree.

I think this would be fine. It looks like the ```BoundTypeExpression.InferredType``` can be moved to ```BoundLocalDeclaration```. 
Alternatively, we could make ```BoundLocalDeclaration.DeclaredType``` optional, and within ```BoundMultipleLocalDeclarations``` set it only for the first ```BoundLocalDeclaration```. The ```BoundTypeExpression.InferredType``` should be moved in this case too. This approach would require less changes in the bound tree structure and you might find that attractive.

> If you agree this should be done, I will obviously do so in a seperate PR as part of a seperate issue.

I think it would be fine to make this change in context of this PR because it is blocked by this issue. It would be fine to do this in a separate PR, but then we will have to wait for it to go through and then continue this PR on top of that work.

Reply:
I tried my suggested solution first. Whilst I preferred this from a purity perspective, and it also simplifies the binding code in a number of situations, it leads to issues in error recovery for cases like this:

```csharp
public class C {
    public void M() {
        var a = 14, b = 78.3; //error CS0819: Implicitly-typed variables cannot have multiple declarators
        
        int c = b; //error CS0266: Cannot implicitly convert type 'double' to 'int'. An explicit conversion exists (are you missing a cast?)
        int d = a;
    }
}
```

https://sharplab.io/#v2:C4LglgNgPgAgzAAhgJgQYQQbwLACgEFKIwAsCAsgBQCUWehDCAbgIYBOCLCAvAgIwkANAgBGPBAHYAHADo4AbnqMCS5WAB2wBAGNxIxfmUINWgCbiWBhgF881oA=

Currently a is inferred as int, and b as double. Hence `int c = b` would give us an error, whilst int d = a would not;.

However this change would make that impossible, as type would now be per declaration statement, not per declaration.

So I'm going to try your approach.

Reply:
>However this change would make that impossible, as type would now be per declaration statement, not per declaration.

>So I'm going to try your approach.

I am pretty sure that with either approach we could keep the same types (as the types before the change) for the locals, but I let you to decide what approach to use. 


---
In reply to: [253298803](https://github.com/dotnet/roslyn/pull/32843#discussion_r253298803) [](ancestors = 253298803)